src/neuroconv/basedatainterface.py

```diff
@@ -4,25 +4,36 @@
 from typing import Optional
 
 from pynwb import NWBFile
+from pydantic.schema import model_schema
+from pydantic.decorator import ValidatedFunction
 
 from .utils import get_base_schema, get_schema_from_method_signature
 
 
 class BaseDataInterface(ABC):
     """Abstract class defining the structure of all DataInterfaces."""
 
+    @classmethod
+    def get_source_model(cls):
+        """Infer the Pydantic model for the source_data from the method signature (annotation typing)."""
+        return ValidatedFunction(function=cls.__init__, config=None).model
+
     @classmethod
     def get_source_schema(cls):
         """Infer the JSON schema for the source_data from the method signature (annotation typing)."""
-        return get_schema_from_method_signature(cls.__init__, exclude=["source_data"])
+        return model_schema(cls.get_source_model())
 
     @classmethod
     def get_conversion_options_schema(cls):
         """Infer the JSON schema for the conversion options from the method signature (annotation typing)."""
         return get_schema_from_method_signature(cls.run_conversion, exclude=["nwbfile", "metadata"])
 
+    def _validate_source(self, source_data: dict):
+        self.get_source_schema().parse_obj(obj=source_data)
```

Comment:
This will also give more informative validation errors - currently, if there's a big `source_data` structure for a converter, it will report a validation error on the entire structure which can sometimes be so big as to completely fill the screen/console - more an issue with `metadata` than with `source`, but still conceptually possible